### PR TITLE
[Merged by Bors] - Fix node update

### LINKED ISSUE #3784: `Changed<Node>` is triggered on each frame
## Bevy version

0.6.0

## What you did

Apply the following patch to bevy source tree (git apply patch.txt)
```rust
diff --git a/examples/ui/button.rs b/examples/ui/button.rs
index 48e67b55..a526f854 100644
--- a/examples/ui/button.rs
+++ b/examples/ui/button.rs
@@ -7,6 +7,7 @@ fn main() {
         .add_plugins(DefaultPlugins)
         .add_startup_system(setup)
         .add_system(button_system)
+        .add_system(ui_size)
         .run();
 }
 
@@ -73,3 +74,9 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
             });
         });
 }
+
+pub fn ui_size(q: Query<&Node, Changed<Node>>) {
+    for node in q.iter() {
+        bevy::log::warn!("Node size change: {:?}", node);
+    }
+}
```

And run `cargo run --example button`

## What you expected to happen

Should print a line only when the UI size actually changes (e.g. when resizing a window).

## What actually happened

This will print UI Node change on each frame:
```
2022-01-27T20:30:42.762507Z  WARN button: Node size change: Node { size: Vec2(150.0, 65.0) }
2022-01-27T20:30:42.778312Z  WARN button: Node size change: Node { size: Vec2(104.0, 40.0) }
2022-01-27T20:30:42.778373Z  WARN button: Node size change: Node { size: Vec2(150.0, 65.0) }
2022-01-27T20:30:42.796493Z  WARN button: Node size change: Node { size: Vec2(104.0, 40.0) }
```
(note the two separate entities with different sizes)

## Additional information

N/A

### PR DESCRIPTION
**Pull request by blaind**
# Objective

Fixes #3784

## Solution

Check if the node size is actually different from previous

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Solid fix. It looks like we can't use change detection here, because we're reliant on information from outside the entity.

File: crates/bevy_ui/src/flex/mod.rs [hunk 1]
```diff
@@ -278,10 +278,13 @@ pub fn flex_node_system(
     // PERF: try doing this incrementally
     for (entity, mut node, mut transform, parent) in node_transform_query.iter_mut() {
         let layout = flex_surface.get_layout(entity).unwrap();
-        node.size = Vec2::new(
+        let new_size = Vec2::new(
             to_logical(layout.size.width),
             to_logical(layout.size.height),
         );
+        if node.size != new_size {
```
Comment (james7132) [inline crates/bevy_ui/src/flex/mod.rs:285]: Potentially leave a comment here saying why this is done.

**Review by james7132 — APPROVED** [review]
Generally looks good. We should probably audit more cases like this where components are being deref'ed mutably without actually changing anything.

**Review by mockersf — APPROVED** [review]

**Pull request closed without merging**